common/ethereum/applications/chronos/chronos.go

```diff
@@ -0,0 +1,341 @@
+		)
+	}
+
+	token1addr_, err := ethereum.StaticCall(client, contractAddr, chronosPairAbi, "token1")
+
+	if err != nil {
+		return nil, fmt.Errorf(
+			"Failed to get token1 address! %v",
+			err,
+		)
+	}
+
+	token1addr, err := ethereum.CoerceBoundContractResultsToAddress(token1addr_)
+
+	if err != nil {
+		return nil, fmt.Errorf(
+			"Failed to coerce token1 address! %v",
+			err,
+		)
+	}
+
+	var (
+		// swap logs
+		amount0in  = swapAmounts[0]
+		amount1in  = swapAmounts[1]
+		amount0out = swapAmounts[2]
+		amount1out = swapAmounts[3]
+
+		// the amount of fluid tokens sent or received
+		fluidTransferAmount *big.Rat
+
+		// the multiplier to find the fee
+		feeMultiplier *big.Rat
+
+		// whether the token being swapped from is the fluid token
+		inTokenIsFluid = false
+	)
+
+	// fluid index 0  0, x, y, 0 = swapping x token for y fluid
+	// fluid index 0  x, 0, 0, y = swapping x fluid for y token
+	// fluid index 1  0, x, y, 0 = swapping x fluid for y token
+	// fluid index 1  x, 0, 0, y = swapping x token for y fluid
+
+	var (
+		// Whether token0 is the fluid token
+		fluidIndex0 = token0addr == fluidTokenContract
+		// Whether swap contains any fluid tokens
+		swapContainsFluid = fluidIndex0 || (token1addr == fluidTokenContract)
+		// Whether amount0 is equal to zero
+		amount0IsZero = amount0in.Sign() == 0
+
+	)
+
+
+	switch true {
+	case !swapContainsFluid:
+		log.App(func(k *log.Log) {
+			k.Format(
+				"Received a Chronos swap in transaction %#v not involving the fluid token - skipping!",
+				transfer.TransactionHash.String(),
+			)
+		})
+
+		return nil, nil
+
+	case fluidIndex0 && amount0IsZero:
+		inTokenIsFluid = false
+		fluidTransferAmount = amount0out
+
+	case fluidIndex0 && !amount0IsZero:
+		inTokenIsFluid = true
+		fluidTransferAmount = amount0in
+
+	case !fluidIndex0 && amount0IsZero:
+		inTokenIsFluid = true
+		fluidTransferAmount = amount1in
+
+	case !fluidIndex0 && !amount0IsZero:
+		inTokenIsFluid = false
+		fluidTransferAmount = amount1out
+	}
+
+	// figure out whether our pair is stable
+	stable_, err := ethereum.StaticCall(client, contractAddr, chronosPairAbi, "isStable")
+
+	if err != nil {
+		return nil, fmt.Errorf(
+			"Failed to get pair isStable! %v",
+			err,
+		)
+	}
+
+	stable, err := ethereum.CoerceBoundContractResultsToBool(stable_)
+
+	if err != nil {
+		return nil, fmt.Errorf(
+			"Failed to coerce isStable bool! %v",
+			err,
+		)
+	}
+
+	// numerator of the fee (must be a call to the pair contract with the token in)
+	feeNumerator_, err := ethereum.StaticCall(client, factoryAddr, chronosPairFactoryAbi, "getFee", stable)
+
+	if err != nil {
+		return nil, fmt.Errorf(
+			"Failed to get pair fee numerator! %v",
+			err,
+		)
+	}
+
+	feeNumerator, err := ethereum.CoerceBoundContractResultsToInt(feeNumerator_)
+
+	if err != nil {
+		return nil, fmt.Errorf(
+			"Failed to coerce pair fee numerator! %v",
+			err,
+		)
+	}
+
+	feeMultiplier = new(big.Rat).SetFrac(feeNumerator, big.NewInt(FeeDenominator))
```

Comment:
What's this for?

Reply:
getting the fee% with correct numerator/denominator